### PR TITLE
Azure Storage Block Blob Client Sync Stack

### PR DESCRIPTION
**Pull request by g2vinay**
Regenerates Storage Blob SDK with Sync Stack and wires it into the convenience layer for following Storage Clients:
* Block Blob Client

This work is being done as part of this branching strategy:
image

### CODE REVIEW

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlockBlobClient.java [hunk 1]
```diff
@@ -1088,9 +1092,51 @@ public Response<BlockBlobItem> commitBlockListWithResponse(List<String> base64Bl
...
+        if (context == null || context == Context.NONE) {
+            return STATIC_ENABLE_REST_PROXY_CONTEXT;
+        } else {
+            return context.addData(HTTP_REST_PROXY_SYNC_PROXY_ENABLE, true);
+        }
+    }
+
+
+
+    <T> Response<T> executeOperation(Supplier<Response<T>> operation, Duration timeout) {
```
Comment (g2vinay) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlockBlobClient.java:1109]: Refactor to StorageImplUtils

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlockBlobClient.java [hunk 2]
```diff
@@ -1088,9 +1092,51 @@ public Response<BlockBlobItem> commitBlockListWithResponse(List<String> base64Bl
...
+        }
+    }
+
+    static ExecutorService getThreadPoolWithShutdownHook() {
+        ExecutorService threadPool = Executors.newCachedThreadPool();
+        registerShutdownHook(threadPool);
+        return threadPool;
+    }
+
+    static Thread registerShutdownHook(ExecutorService threadPool) {
```
Comment (g2vinay) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlockBlobClient.java:1125]: Refactor to StorageImplUtils

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlockBlobClient.java [hunk 3]
```diff
@@ -1088,9 +1092,51 @@ public Response<BlockBlobItem> commitBlockListWithResponse(List<String> base64Bl
...
     public Response<BlockBlobItem> commitBlockListWithResponse(BlockBlobCommitBlockListOptions options,
         Duration timeout, Context context) {
-        Mono<Response<BlockBlobItem>> response = client.commitBlockListWithResponse(
-            options, context);
+        return executeOperation(() -> client.commitBlockListWithResponseSync(options, enableSyncRestProxy(context)),
+            timeout);
+    }
 
-        return blockWithOptionalTimeout(response, timeout);
+    public static Context enableSyncRestProxy(Context context) {
```
Comment (g2vinay) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlockBlobClient.java:1099]: Refactor to Storage Impl Utils

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlockBlobClient.java [hunk 4]
```diff
@@ -32,22 +33,18 @@
...
-
 import java.io.InputStream;
 import java.io.UncheckedIOException;
 import java.net.URL;
-import java.nio.ByteBuffer;
 import java.time.Duration;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
+import java.util.concurrent.*;
```
Comment (g2vinay) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlockBlobClient.java:45]: fix checkstyle

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlockBlobAsyncClient.java [hunk 5]
```diff
@@ -448,6 +452,39 @@ Mono<Response<BlockBlobItem>> uploadWithResponse(BlockBlobSimpleUploadOptions op
             }));
     }
 
+    Response<BlockBlobItem> uploadWithResponseSync(BlockBlobSimpleUploadOptions options, Context context) {
```
[the remark this answers is not in the document]
Comment (g2vinay) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlockBlobAsyncClient.java]: moved to sync client with a TODO to move constructor setup to base class once all STG clients support sync stack.

File: sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/implementation/StorageImplUtils.java [hunk 6]
```diff
@@ -418,4 +430,67 @@ public Map.Entry<String, String> next() {
...
+                }
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                threadPool.shutdown();
+            }
+        });
+        Runtime.getRuntime().addShutdownHook(hook);
+    }
+
+    public static String tagsToString(Map<String, String> tags) {
```
Comment (alzimmermsft) [inline sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/implementation/StorageImplUtils.java:475]: Was this copied from elsewhere? This seems familiar. And why was this added?

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlockBlobClient.java [hunk 7]
```diff
@@ -594,12 +748,48 @@ public BlockBlobItem uploadFromUrl(String sourceUrl, boolean overwrite) {
     public Response<BlockBlobItem> uploadFromUrlWithResponse(BlobUploadFromUrlOptions options, Duration timeout,
                                                              Context context) {
         StorageImplUtils.assertNotNull("options", options);
-        Mono<Response<BlockBlobItem>> upload = client.uploadFromUrlWithResponse(options, context);
+        return StorageImplUtils.executeOperation(() -> uploadFromUrlWithResponseSync(options, enableSyncRestProxy(context)),
+            timeout);
+    }
+
+    Response<BlockBlobItem> uploadFromUrlWithResponseSync(BlobUploadFromUrlOptions options, Context context) {
+        StorageImplUtils.assertNotNull("options", options);
```
Comment (alzimmermsft) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlockBlobClient.java:756]: Either this check or the check in `uploadFromUtilWithResponse` can be removed as they are duplicative.

File: sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/implementation/StorageImplUtils.java (hunk 6 above)
Comment (g2vinay) [reply sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/implementation/StorageImplUtils.java:475]: This was copied from async client I believe when adding sync code to sync client.
Updated async client to reference it from here.
We can move it back to clients too and have a static package private copy there, if its not needed here.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlockBlobClient.java [hunk 8]
```diff
@@ -1088,9 +1320,161 @@ public Response<BlockBlobItem> commitBlockListWithResponse(List<String> base64Bl
...
+            .customerProvidedKey(encryptionKey)
+            .encryptionScope(this.getEncryptionScope());
+    }
+
+    /**
+     * Gets the {@link HttpPipeline} powering this client.
+     *
+     * @return The pipeline.
+     */
+    public HttpPipeline getHttpPipeline() {
```
[the remark this answers is not in the document]
Comment (g2vinay) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlockBlobClient.java:1418]: Yeah the goal was to override the parent class behaviour of invoking async client to fetch the data.
The getters return local stored variables.
